edge-util/src/Microsoft.Azure.Devices.Edge.Storage/DbStoreProviderWithBackupRestore.cs

```diff
@@ -0,0 +1,351 @@
+        readonly IDbStoreBackupRestore dbStoreBackupRestore;
+        readonly SerializationFormat backupFormat;
+        readonly Events events;
+
+        DbStoreProviderWithBackupRestore(
+            IDbStoreProvider dbStoreProvider,
+            string backupPath,
+            IDbStoreBackupRestore dbStoreBackupRestore,
+            SerializationFormat backupFormat)
+            : base(dbStoreProvider)
+        {
+            this.backupPath = Preconditions.CheckNonWhiteSpace(backupPath, nameof(backupPath));
+            this.dbStoreBackupRestore = Preconditions.CheckNotNull(dbStoreBackupRestore, nameof(dbStoreBackupRestore));
+            this.backupFormat = backupFormat;
+            this.dbStores = new HashSet<string>();
+            this.events = new Events(this.Log);
+        }
+
+        public static async Task<IDbStoreProvider> CreateAsync(
+            IDbStoreProvider dbStoreProvider,
+            string backupPath,
+            IDbStoreBackupRestore dbStoreBackupRestore,
+            SerializationFormat backupFormat)
+        {
+            DbStoreProviderWithBackupRestore provider = new DbStoreProviderWithBackupRestore(dbStoreProvider, backupPath, dbStoreBackupRestore, backupFormat);
+            await provider.RestoreAsync();
+            return provider;
+        }
+
+        private async Task RestoreAsync()
+        {
+            string backupMetadataFilePath = Path.Combine(this.backupPath, BackupMetadataFileName);
+
+            if (!File.Exists(backupMetadataFilePath))
+            {
+                this.events.NoBackupsForRestore();
+                return;
+            }
+
+            try
+            {
+                string fileText = File.ReadAllText(backupMetadataFilePath);
+                BackupMetadataList backupMetadataList = JsonConvert.DeserializeObject<BackupMetadataList>(fileText);
+                BackupMetadata backupMetadata = backupMetadataList.Backups[0];
+                this.events.BackupInformation(backupMetadata.Id, backupMetadata.SerializationFormat, backupMetadata.TimestampUtc, backupMetadata.Stores);
+
+                string latestBackupDirPath = Path.Combine(this.backupPath, backupMetadata.Id.ToString());
+                if (Directory.Exists(latestBackupDirPath))
+                {
+                    this.events.RestoringFromBackup(backupMetadata.Id);
+                    foreach (string store in backupMetadata.Stores)
+                    {
+                        IDbStore dbStore;
+                        if (!store.Equals(DefaultStoreBackupName, StringComparison.OrdinalIgnoreCase))
+                        {
+                            dbStore = this.GetDbStore(store);
+                        }
+                        else
+                        {
+                            dbStore = this.GetDbStore();
+                        }
+
+                        await this.dbStoreBackupRestore.RestoreAsync(store, dbStore, latestBackupDirPath);
+                    }
```

Comment:
I think it should not pass the dbStore and add items to it in the RestoreAsync, but return the items that were restored and the code that iterates over the items to add them to be moved here

Reply:
Removed DbStoreBackupRestore and moved it's code to DbStoreProviderWithBackupRestore as per our discussion